File: leetcode/solutions/0876.MiddleOfTheLinkedList/middleNode.ts
import { SinglyLinkedListNode } from "../../../datastructure/SinglyLinkedListNode.ts";

export default function middleNode(
  p1: SinglyLinkedListNode<number> | null
): SinglyLinkedListNode<number> | null {
  if (!p1) {
    return p1;
  }

  let p2: SinglyLinkedListNode<number> | null = p1;

  while (p2?.next) {
    p1 = p1!.next;
    p2 = p2.next.next;
  }

  return p1;
}

File: leetcode/solutions/0876.MiddleOfTheLinkedList/middleNode.go
package solution

import (
	"github.com/yukisato/leetcode-util/datastructure"
)

// ListNode structure for LeetCode
type ListNode = datastructure.ListNode

func middleNode(head *ListNode) *ListNode {
	result := head

	for head != nil && head.Next != nil {
		head = head.Next.Next
		result = result.Next
	}

	return result
}
